README.md
# Welcome to DB UI Elements

![Apache 2.0 license badge](https://img.shields.io/badge/License-Apache_2.0-blue.svg)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](https://makeapullrequest.com)

DB UI Elements is a library, that provides style conform UI components for Deutsche Bahn web
frontends. It's designed to work with any web framework — or even without one.

## How to consume

Find a here a detailed [how to consume](doc/howto-start.md) for your development context, either vanilla JavaScript or any framework – we're proving "native" framework components for Angular, React and Vue.

## Core principals

### Framework agnostic

DB UI Elements is built on native Web Standards and can be used in any web framework like (Angular, React, Vuejs) or even without one. The underlying technology is Web Components, which is a set of native Web Standards that work together, letting you create custom elements that work just like the standard HTML elements built into your browser.

### Compliant to Deutsche Bahn UI/UX guidelines

DB UI Elements follows the Deutsche Bahn UI/UX guidelines.
[See more details Styling](/doc/howto-style.md)

### Multi-tenant

Every tenant in Deutsche Bahn has a deviation from corporate style.  
We separate style from implementation. So that the consumer can decide, which tenant style to support.
It is also possible to use multiple styles in the same application, just by switching the configuration.

The style is provided by [DB UI Core](https://github.com/db-ui/core)

Currently we support following tenants:

* Regular (DB UX Design System)
* Enterprise

Other tenants can be added -> please contact us

### Accessible

Web Components are a set of technologies that work together, letting you create custom elements that work just like the standard HTML elements built into your browser.

### Light weight

DB UI Elements Web Components are implemented using StencilJS, which is designed for creating web components with a minimum amount of overhead.

### Open and free to use

DB UI Elements is free (Apache 2.0 license) to consume and is community driven.
If you miss a component than [contribute](CONTRIBUTING.md), please.

## What is next?

[Overview on documentation](doc/README.md)

## Framework showcases

* [DB UI Elements with React (CRA)](showcase/react-showcase/README.md)
* [DB UI Elements with Angular](showcase/angular-lts-showcase/README.md)
* [DB UI Elements with Vue](showcase/vue-showcase/README.md)

## Deutsche Bahn brand

As we'd like to perfectly support our users and customers on their digital journey, the usage of Deutsche Bahn brand and trademarks are bound to clear guidelines and restrictions even when being used with the code that we're providing with this product.
Deutsche Bahn fully reserves all rights and ownership regarding the Deutsche Bahn brand, even though that we're providing the code of DB UI products free to use and release it under the Apache 2.0 license.
Please have a look at our [brand portal](https://marketingportal.extranet.deutschebahn.com/) for any further questions and whom to contact on any brand issues.

As these assets and visual guidelines are retrieved from our Deutsche Bahn Marketingportal, you'll agree with the ["Allgemeine Nutzungsbedingungen für das DB-Marketingportal" (german)](https://marketingportal.extranet.deutschebahn.com/marketingportal/Nutzungsbedingungen-9702684) in case of using them.

For any usage outside of Deutsche Bahn websites and applications you must remove or replace any Deutsche Bahn brand and design assets as well as protected characteristics and trademarks. We're even also planning to provide a neutral theme that would make it much easier for you to use our product without the trademarks by Deutsche Bahn.

This especially relates to (but doesn't exclude further Deutsche Bahn brand assets) the file `db_logo.svg` that is duplicated in several different places.

## Contributions

Contributions are very welcome, please refer to the [contribution guide](CONTRIBUTING.md).

## Third party acknowledgments – many kudos for that !!!

- [open privacy by opr.vc](https://opr.vc)

## Code of conduct

We as members, contributors, and leaders pledge to make participation in our
community a harassment-free experience for everyone – have a look at our [Contributor Covenant Code of Conduct](CODE-OF-CONDUCT.md).

## License

This project is licensed under [Apache-2.0](LICENSE), Copyright 2022 by DB Systel GmbH.


CHANGELOG.md
<!-- markdownlint-configure-file { "MD024": false, "MD013": false } -->

# Changelog

| :point_up: | From now on we'll only document our changes within the [repositories releases page](https://github.com/db-ui/core/releases). |
| ---------- | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

~All notable changes to this project will be documented in this file.~

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
And the commit guidelines from [Conventional Commits](https://conventionalcommits.org) are being used.

Regarding BREAKING CHANGES in between, please have a look into [the migration guide](doc/migrationGuide.html).

## [Unreleased]

## [0.14.4] - 2022-07-12

### Changed

#### DEV

- updated `storybook`, `stencil` & DB UI Core dependencies

### Fixed

#### Package consumption

- Correctly displaying the icons within buttons

## [0.14.3] - 2022-07-06

### Fixed

#### Package consumption

- Correctly displaying the icons within buttons

## [0.14.2] - 2022-06-15

### Added

#### Package consumption

- `LICENSE` files to all packages

#### DEV

- Playground: Visual page builder

### Changed

#### Package consumption

- further aspects regarding DB trademarks

#### DEV

- DB UI Core update; a.o. icons path and filenames have changed

### Fixed

#### Package consumption

- Output target for **Vue / Vite consumption**
- Replaced remaining SCSS functions (font calculation e.g. `to-rem`) in CSS files by updating DB UI Core
- Added missing `"license"` field to `package.json` by Vue packages
- Issue with generated sourcemaps for React & Vue output-targets

## [0.13.0] - 2022-04-22

### Added

#### Package consumption

- [`reactwind`](https://db-ui.github.io/elements/?path=/story/intro-showcases--react-and-tailwind) showcase
- Providing a [`developer playground`](https://db-ui.github.io/elements/?path=/story/development-playground--editor)
- `CONTRIBUTING`, `MAINTAINERS` and `CODE-OF-CONDUCT` files

### Changed

#### Package consumption

- **BREAKING CHANGE**: *expander* component got renamed to *accordion* and has some new configurable options
- **BREAKING CHANGE**: The persistant state type of the `tags` component has been extracted into a new `chips` component

#### DEV

- Consolidated showcases
- Updated DB UI Core
- Providing framework-agnostic cypress test cases
- Consolidated `npm` scripts

### Fixed

#### Package consumption

- brand component still contained a non-configurable asset path


#### DEV

- Angular showcase warning output on startup

## [0.12.1] - 2022-04-06

### Fixed

#### Package consumption

- issue with Angular dependencies on Angular 13 [!219]

#### DEV

- Warning message on Vue showcase build regarding `db-textarea` [#92]

## [0.12.0] - 2022-04-05

### Added

#### Package consumption

- feat: added new component `db-card` / `db-cards` [!175] [!32]
- feat: added new component `db-headline` [!212] [#80]
- feat: added new component `db-language-switcher` [!180] [#35]
- feat: added new component `db-mainnavigation` [!173] [#36]
- feat: added new component `db-notification` [!141] [#38]
- feat: added new component `db-pagination` [!176] [#40]
- chore: update angular documentation to `handle inject()` error [!215]

#### DEV

- add coverage for cypress tests [!211]

### Changed

#### DEV

- dependency updates: `np`, `yargs`, `@stencil/core`, `nanoid`, `@db-ui/core`
- optimized the documentation

### Fixed

#### Package consumption

- "Husky Angular Component Check" [!217] [#94]

## [0.11.0] - 2022-03-25

### Added

- update `linklist` and `metanavigation` to work with `<li>` and data

### Fixed

- gitlab pages for master

## [0.10.2] - 2022-03-24

### Changed

- refactor: generalizing the SCSS declarations necessary for every component

## [0.10.1] - 2022-03-24

### Changed

- Optimized the image path within the `db-brand` component (nevertheless we still need to have this configurable)
- Further optimizations within the documentation and examples

### Fixed

- issues with gitlab pages and showcases on `master`
- style: issue with `db-page`

## [0.10.0] - 2022-03-23

### Added

- build: added framework showcases (react, angular, vue) to monorepo
- feat: added new component `db-brand`
- feat: added new component `db-breadcrumb`
- feat: added new component `db-dropdown`
- feat: added new component `db-expander`
- feat: added new component `db-footer`
- feat: added new component `db-header`
- feat: added new component `db-linklist`
- feat: added new component `db-metanavigation`
- feat: added new component `db-overflow-menu`
- feat: added new component `db-sidenavi`
- feat: added new component `db-tab-bar`
- feat: added new component `db-table`


## [0.8.1] - 2021-09-23

### Added

- build: added secrets as a git submodule

### Changed

- refactor: updating nginx and node
- chore(package): update @commitlint dependencies to version 13.1.0
- chore(package): update @storybook dependencies
- chore: switched the `commitlint` config file format
- build(dependencies): update prettier dependency to version 2.4.1
- build(dependencies): update prettier dependency to version 2.4.1

### Fixed

- theme: corrected css references within the themes
- corrected a lot of documentation parts
- refactor: switching the registry URL [#50]

## [0.6.2] - 2020-10-06

### Fixed

- Link: further optimized the horizontal placement of the icons

## [0.6.1] - 2020-10-05

### Fixed

- Checkbox: vertical alignment of the label

## [0.6.0] - 2020-10-05

### Added

- textarea component

### Changed

- Remove custom-elements.json from git as it's a generated file and shouldn't get version controlled

### Fixed

- The input component was missing its styling
- Docs: some corrections especially regarding feedback

.storybook/preview.js
//enable watch mode for hot deployment during development
import { defineCustomElements } from '@db-ui/elements/dist/esm/loader';
defineCustomElements();

// order the Pages top level
export const parameters = {
  actions: { argTypesRegex: '^on[A-Z].*' },
  layout: 'centered',
  controls: { expanded: true, hideNoControlsWarning: true },
  previewTabs: {
    'storybook/docs/panel': { hidden: true },
    canvas: { hidden: true }
  },
  viewMode: 'story',
  options: {
    storySort: {
      method: 'alphabetical',
      order: [
        '00-Intro',
        ['Welcome', 'Getting Started', 'Frameworks', 'Showcases'],
        '01-Layout',
        '02-Navigation',
        '03-Action',
        '04-Data-Display',
        '05-Data-Input',
        '06-Feedback',
        '07-Utility',
        '99-Development'
      ]
    }
  }
};
